packages/vitest/src/node/watch-filter.ts

```diff
@@ -0,0 +1,160 @@
+import readline from 'node:readline'
+import c from 'picocolors'
+import { stdout } from '../utils'
+
+const MAX_RESULT_COUNT = 10
+const SELECTION_MAX_INDEX = 7
+const ESC = '\u001B['
+
+type FilterFunc = (keyword: string) => Promise<string[]> | string[]
+
+export class WatchFilter {
+  private filterRL: readline.Interface
+  private currentKeyword: string | undefined = undefined
+  private message: string
+  private results: string[] = []
+  private selectionIndex = -1
+  private onKeyPress?: (str: string, key: any) => void
+
+  constructor(message: string) {
+    this.message = message
+    this.filterRL = readline.createInterface({ input: process.stdin, escapeCodeTimeout: 50 })
+    readline.emitKeypressEvents(process.stdin, this.filterRL)
+    if (process.stdin.isTTY)
+      process.stdin.setRawMode(true)
+  }
+
+  public async filter(filterFunc: FilterFunc): Promise<string | undefined> {
+    stdout().write(`${c.cyan('?')} ${c.bold(this.message)} › `)
+    return new Promise((resolve) => {
+      this.onKeyPress = this.filterHandler(filterFunc, (result) => {
+        this.close()
+        resolve(result)
+      })
+      process.stdin.on('keypress', this.onKeyPress)
+    })
```

Comment:
(refactoring suggestion) `@vitest/utils` has `createDefer`, which makes `new Promise` look less convoluted to my eye. I tend to prefer this pattern. What do you think?

```tsx
    const resultPromise = createDefer<string | undefined>();
    this.onKeyPress = this.filterHandler(filterFunc, (result) => {
      resultPromise.resolve(result);
    })
    process.stdin.on('keypress', this.onKeyPress)
    try {
      return await resultPromise;
    } finally {
      this.close()
    }
```

Reply:
@hi-ogawa 
Thank you!
I've refactored my code to use createDefer [3934ae9](https://github.com/vitest-dev/vitest/pull/4733/commits/3934ae957250e75d49b342584dbab45c61dd6f14)